Code/EnginePlugins/OpenXRPlugin/OpenXRSingleton.cpp

```diff
@@ -0,0 +1,1122 @@
+        ezMath::Tan(ezAngle::Radian(view.fov.angleLeft)) * cam->GetNearPlane(),
+        ezMath::Tan(ezAngle::Radian(view.fov.angleRight)) * cam->GetNearPlane(),
+        ezMath::Tan(ezAngle::Radian(view.fov.angleDown)) * cam->GetNearPlane(),
+        ezMath::Tan(ezAngle::Radian(view.fov.angleUp)) * cam->GetNearPlane(), cam->GetNearPlane(), cam->GetFarPlane());
+    };
+
+    // Update projection with newest near/ far values. If not sync camera is set, just use the last value from XR
+    // camera.
+    const ezMat4 projLeft = CreateProjection(m_views[0], m_pCameraToSynchronize);
+    const ezMat4 projRight = CreateProjection(m_views[1], m_pCameraToSynchronize);
+    m_pCameraToSynchronize->SetStereoProjection(projLeft, projRight, fAspectRatio);
+    m_uiSettingsModificationCounter = m_pCameraToSynchronize->GetSettingsModificationCounter();
+  }
+
+  // Update camera view
+  {
+    ezTransform add;
+    add.SetIdentity();
+    ezView* pView = nullptr;
+    if (ezRenderWorld::TryGetView(m_hView, pView))
+    {
+      if (const ezWorld* pWorld = pView->GetWorld())
+      {
+        EZ_LOCK(pWorld->GetReadMarker());
+        if (const ezStageSpaceComponentManager* pStageMan = pWorld->GetComponentManager<ezStageSpaceComponentManager>())
+        {
+          if (const ezStageSpaceComponent* pStage = pStageMan->GetSingletonComponent())
+          {
+            ezEnum<ezXRStageSpace> stageSpace = pStage->GetStageSpace();
+            if (m_StageSpace != stageSpace)
+              SetStageSpace(pStage->GetStageSpace());
+            add = pStage->GetOwner()->GetGlobalTransform();
+          }
+        }
+      }
+    }
+
+    {
+      // Update device state (average of both eyes).
+      ezQuat rot;
+      rot.SetIdentity();
+      rot.SetSlerp(
+        ConvertOrientation(m_views[0].pose.orientation), ConvertOrientation(m_views[1].pose.orientation), 0.5f);
+      const ezVec3 pos =
+        ezMath::Lerp(ConvertPosition(m_views[0].pose.position), ConvertPosition(m_views[1].pose.position), 0.5f);
+
+      m_Input->m_DeviceState[0].m_vGripPosition = pos;
+      m_Input->m_DeviceState[0].m_qGripRotation = rot;
+      m_Input->m_DeviceState[0].m_Type = ezXRDeviceType::HMD;
+      m_Input->m_DeviceState[0].m_bGripPoseIsValid = true;
+      m_Input->m_DeviceState[0].m_bDeviceIsConnected = true;
+    }
+
+    // Set view matrix
+    {
+      const ezMat4 mStageTransform = add.GetAsMat4();
+      const ezMat4 poseLeft = mStageTransform * ConvertPoseToMatrix(m_views[0].pose);
+      const ezMat4 poseRight = mStageTransform * ConvertPoseToMatrix(m_views[1].pose);
+
+      // EZ Forward is +X, need to add this to align the forward projection
+      const ezMat4 viewMatrix =
+        ezGraphicsUtils::CreateLookAtViewMatrix(ezVec3::ZeroVector(), ezVec3(1, 0, 0), ezVec3(0, 0, 1));
+      const ezMat4 mViewTransformLeft = viewMatrix * poseLeft.GetInverse();
+      const ezMat4 mViewTransformRight = viewMatrix * poseRight.GetInverse();
```

Comment:
This may depend on the coordinate system configuration, which is set in the ezWorld. @C-Core should this be handled here?

Reply:
Yes, if you want to make this future proof, get the coordinate system provider from the world to obtain the forward axis etc

---
In reply to: [423524360](https://github.com/ezEngine/ezEngine/pull/212#discussion_r423524360) [](ancestors = 423524360)

Reply:
Ok I'll put that on my list of things to improve. Too big of a task to do in this PR but might be a good idea to do that before reviving the OpenVR plugin,

Reply:
Frankly, I wouldn't mind ignoring this for the time being. Was just pointing it out, but in practice it's not a problem.